packages/block-events-stream/src/createBlockStream.ts

```diff
@@ -0,0 +1,38 @@
+import { BehaviorSubject } from "rxjs";
+import type { Block, BlockTag, PublicClient } from "viem";
+import { ReadonlyBehaviorSubject } from "./common";
+
+// TODO: pass through viem's types, e.g. WatchBlocksParameters -> GetBlockReturnType
+// TODO: make stream closeable to make use of unwatch?
```

Comment:
this is a good point, i think we have to provide a way to unsubscribe all the internal subscriptions (ie everytime we set up a BehaviorSubject to return and do `.subscribe` somewhere), otherwise we might end up with a lot of subscriptions when the client hot reloads or sth

Reply:
yeah I wasn't sure how best to handle this given we're in the context of a stream

we could create our own type/wrapper around `BehaviorSubject` that has a `close` method or something, which will unsubscribe, but felt weird and error prone to extend rxjs types

we could return both the stream and unsubscribe/close methods, but that also felt weird

Reply:
Agree it's tricky, wonder what others do in this situation. I think we can't get around doing one of those to allow cleaning up old subscriptions (we ran into a bunch of issues with duplicate subscriptions in the past due to not cleaning them up properly, that was one of the reasons `recs`'s `world` had the concept of `disposers`). Returning both the stream and an `unsubscribe` method feels a bit better than extending the `BehaviorSubject` class because it makes it more explicit that this is something that should be unsubscribed. 

Btw I think if all of these were just `Observables` instead of `Subjects` only the last subscriber would have to unsubscribe, but it comes with its own set of issues (no concept of "current" value, issues with multiple subscribers to the same observable, and we would still need to pass through things like the `watchBlocks` `close` method)

Reply:
worth also deciding if rxjs is what we want as the underlying primitive vs. a native JS stream or event emitter or some other thing

Reply:
After our chat, I did some digging and native streams are quite clunky. They only support byte buffers (so that the streams can move between threads), so we'd be JSON serializing and deserializing a bunch. They're kind of a pain to construct and read/write from. Since it's a pure byte stream, I don't think we can guarantee JSON blobs would come in meaningful chunks that we could immediately parse, so we'd need some separator between each JSON object and a stream parser that is JSON aware to correctly split on those separators (and not if the separator exists in a JSON string inside an object).

I couldn't find any new libs that wrap native streams to make them easier to use. A bunch of the libs I found are meant for Node.js streams, which deviate a bit from the browser stream spec (like `objectMode`).